src/test/java/io/lettuce/core/cluster/pubsub/RedisClusterPubSubConnectionIntegrationTests.java

```diff
@@ -103,36 +96,41 @@ void testRegularClientPubSubChannels() {
     @Test
     void testRegularClientPubSubShardChannels() {
 
-        String nodeId = pubSubConnection.sync().clusterMyId();
-        RedisClusterNode otherNode = getOtherThan(nodeId);
-        /// TODO : uncomment after SSUBSCRIBE is implemented
-        // pubSubConnection.sync().ssubscribe(key);
+        pubSubConnection.sync().ssubscribe(shardChannel);
+
+        Integer clusterKeyslot = connection.sync().clusterKeyslot(shardChannel).intValue();
+        RedisCommands<String, String> rightSlot =
+                connection.sync().nodes(node -> node.getSlots().contains(clusterKeyslot)).commands(0);
+        RedisCommands<String, String> wrongSlot =
+                connection.sync().nodes(node -> !node.getSlots().contains(clusterKeyslot)).commands(0);
 
-        List<String> channelsOnSubscribedNode = connection.getConnection(nodeId).sync().pubsubShardChannels();
-        // assertThat(channelsOnSubscribedNode).hasSize(1);
+        List<String> channelsOnSubscribedNode = rightSlot.pubsubShardChannels();
+        assertThat(channelsOnSubscribedNode).hasSize(1);
 
-        List<String> channelsOnOtherNode = connection.getConnection(otherNode.getNodeId()).sync().pubsubShardChannels();
+        List<String> channelsOnOtherNode = wrongSlot.pubsubShardChannels();
         assertThat(channelsOnOtherNode).isEmpty();
     }
 
     @Test
     @EnabledOnCommand("SSUBSCRIBE")
-    void subscribeToShardChannel() throws Exception {
-        pubSubConnection.addListener(connectionListener);
-        pubSubConnection.async().ssubscribe(shardChannel);
-        assertThat(connectionListener.getChannels().poll(3, TimeUnit.SECONDS)).isEqualTo(shardChannel);
+    void subscribeToShardChannel(){
+        pubSubConnection.sync().ssubscribe(shardChannel);
+
+        Wait.untilEquals(1L, connectionListener.getShardCounts()::poll).waitOrTimeout();
+        Wait.untilEquals(shardChannel, connectionListener.getShardChannels()::poll).waitOrTimeout();
     }
 
     @Test
     @EnabledOnCommand("SSUBSCRIBE")
-    void subscribeToShardChannelViaOtherEndpoint() throws Exception {
-        pubSubConnection.addListener(connectionListener);
-        RedisClusterPubSubAsyncCommands<String, String> pubSub = pubSubConnection.async();
-        String nodeId = pubSub.clusterMyId().get();
-        RedisPubSubAsyncCommands<String, String> other = pubSub
-                .nodes(node -> node.getRole().isUpstream() && !node.getNodeId().equals(nodeId)).commands(0);
-        other.ssubscribe(shardChannel);
-        assertThat(connectionListener.getChannels().poll(3, TimeUnit.SECONDS)).isEqualTo(shardChannel);
+    void subscribeToShardChannelViaReplica() {
```

Comment:
this was not the initial test case i had in mind, is it possible we extend it to include the other endpoint with a proper channel name as well ? it is just to make sure everything works as expected in case subscribe lands on another node which handles different hash slots.

Reply:
As agreed offline let's keep the implementation as it is.
When @atakavci is done with the SPUBLISH command he would revise the test and suggest a new approach